tensorboard/plugins/custom_scalar/custom_scalars_plugin.py

```diff
@@ -0,0 +1,230 @@
+from tensorboard.plugins.scalar import metadata as scalars_metadata
+from tensorboard.plugins.scalar import scalars_plugin
+
+# The name of the property in the response for whether the regex is valid.
+_REGEX_VALID_PROPERTY = 'regex_valid'
+
+# The name of the property in the response for the payload (tag to ScalarEvents
+# mapping).
+_TAG_TO_EVENTS_PROPERTY = 'tag_to_events'
+
+# The number of seconds to wait in between checks for the config file specifying
+# layout.
+_CONFIG_FILE_CHECK_THROTTLE = 60
+
+
+class CustomScalarsPlugin(base_plugin.TBPlugin):
+  """CustomScalars Plugin for TensorBoard."""
+
+  plugin_name = 'custom_scalars'
+
+  def __init__(self, context):
+    """Instantiates ScalarsPlugin via TensorBoard core.
+
+    Args:
+      context: A base_plugin.TBContext instance.
+    """
+    self._logdir = context.logdir
+    self._multiplexer = context.multiplexer
+    self._scalars_plugin = scalars_plugin.ScalarsPlugin(context)
+
+    # A JSON representation of the layout of the custom scalars dashboard.
+    self._layout = None
+
+    # Start a separate thread that periodically checks for the config file
+    # specifying layout.
+    threading.Thread(target=self._periodically_check_for_config_file).start()
+
+  def get_plugin_apps(self):
+    return {
+        '/download_data': self.download_data_route,
+        '/layout': self.layout_route,
+        '/scalars': self.scalars_route,
+    }
+
+  def is_active(self):
+    """This plugin is active if 2 conditions hold.
+
+    1. The scalars plugin is active.
+    2. There is a custom layout for the dashboard.
+    """
+    return self._scalars_plugin.is_active() and self._layout is not None
+
+  def _periodically_check_for_config_file(self):
+    """Periodically checks disk for the presence on a config file.
+
+    This method is to be run on a separate thread.
+    """
+    try:
+      self._check_for_config_file()
+    except IOError as e:
+      # Log the error. Try to read the config file again later.
+      tf.logging.error(
+          'Failed to read %r from the logdir: %r',
+          layout.CONFIG_FILE, e)
+
+    # Wait a while before checking again.
+    time.sleep(_CONFIG_FILE_CHECK_THROTTLE)
+
+  def _check_for_config_file(self):
+    """Tries to load the config file that specifies layout.
+
+    Raises: IOError if reading the file from disk triggers an error. No
+      exception will be thrown if the file does not exist however.
+    """
+    config_fpath = os.path.join(self._logdir, layout.CONFIG_FILE)
+
+    try:
+      if not tf.gfile.Exists(config_fpath):
+        # The file does not exist right now.
+        return
+    except OSError as e:
+      raise IOError(e)
+
+    try:
+      read_file = tf.gfile.Open(config_fpath, 'r')
+      pbtxt_contents = read_file.read()
+    except OSError as e:
+      raise IOError(e)
+
+    layout_proto = scalars_layout_pb2.Layout()
+    text_format.Parse(pbtxt_contents, layout_proto)
+    self._layout = json_format.MessageToJson(
```

Comment:
I think while in Python-land it should be stored as a proto, not JSON. Early conversion to JSON just loses inspectability/safety, no?

Then, you can convert to JSON in the handler when it is actually being sent to the frontend.

It gives a straightforward mapping between data storage type and location:

fronted = json
backend = proto